### ISSUE TITLE
[Bug] Codellama 34B cannot handle inputs > 4k tokens despite context window of 16k

### ISSUE THREAD

**Issue by nverke** [mlc-ai/mlc-llm#923]
## 🐛 Bug

Codellama 34B model does not behave as expected when passing inputs with more than 4k tokens. The model seems to lose the context and start spitting out random code imports.

Notes:
Codellama 13B does not exhibit the same behavior and can handle inputs up to 16K tokens fine. 
Codellama 34B fp16, int8, and int4 all have the same issue with tokens >4K.

## To Reproduce

1. Download Codellama 34B
2. Compile the model using MLC-LLM 
3. Create a request with >4k tokens
4. Run the model with the request.

## Expected behavior

Model should respect the context of the input and respond accordingly.

## Environment

 - Platform (e.g. WebGPU/Vulkan/IOS/Android/CUDA): CUDA
 - Operating system (e.g. Ubuntu/Windows/MacOS/...): Ubuntu
 - Device (e.g. iPhone 12 Pro, PC+RTX 3090, ...) A100 
 - How you installed MLC-LLM (`conda`, source): Yes
 - How you installed TVM-Unity (`pip`, source): Yes
 - Python version (e.g. 3.10): 3.10
 - GPU driver version (if applicable):
 - CUDA/cuDNN version (if applicable): CUDA 12.0
 - TVM Unity Hash Tag (`python -c "import tvm; print('\n'.join(f'{k}: {v}' for k, v in tvm.support.libinfo().items()))"`, applicable if you compile models): 631f37b6bf8b101d16ecc55de7e6a749a3588570
 - Any other relevant information:

**Comment by MasterJH5574** [comment]
Thanks @nverke for reporting! I seemed to have reproduced the issue. On my side the 34B model errors when the input has length around 6k tokens. We will dig into this.

**Comment by nverke** [comment]
Hey, @MasterJH5574 any chance you can share some more details of the investigation? Any inputs you have used to test?

**Comment by junrushao** [comment]
@nverke @MasterJH5574 Is it even theoretically possible to extend to unlimited context length without proper RoPE scaling at all? I'm not sure about CodeLlama, but Llama2 config seems to suggest `max_position_embeddings = 4096` is the uplimit:

```JSON
{
  "_name_or_path": "meta-llama/Llama-2-70b-chat-hf",
  "architectures": [
    "LlamaForCausalLM"
  ],
  "bos_token_id": 1,
  "eos_token_id": 2,
  "hidden_act": "silu",
  "hidden_size": 8192,
  "initializer_range": 0.02,
  "intermediate_size": 28672,
  "max_position_embeddings": 4096,
  "model_type": "llama",
  "num_attention_heads": 64,
  "num_hidden_layers": 80,
  "num_key_value_heads": 8,
  "pad_token_id": 0,
  "pretraining_tp": 1,
  "rms_norm_eps": 1e-05,
  "rope_scaling": null,
  "tie_word_embeddings": false,
  "torch_dtype": "float16",
  "transformers_version": "4.31.0.dev0",
  "use_cache": true,
  "vocab_size": 32000
}
```

**Comment by nverke** [comment]
I am not sure if rope_scaling is needed or not but here is the config for codellama. Seems that it should support up to 16k. Additionally seems that transformers can handle up to 16k as well.
https://huggingface.co/codellama/CodeLlama-34b-hf/blob/main/config.json

**Comment by junrushao** [comment]
> "max_position_embeddings": 16384

**Comment by MasterJH5574** [comment]
> Any chance you can share some more details of the investigation? Any inputs you have used to test?

Hi @nverke, on our M2 Ultra Mac, the 34B model generates properly when input has 5768 tokens, and fails when the input has 5897 tokens. We didn't further narrow down the range.

The phenomenon we noticed here is, when the model fails, the logits of the prefill stage are all NAN. We don’t yet know the cause of this issue.

The input we use is an arbitrary text snippet ended with `"Please ignore all the above and write a code to compute the fibonacci sequence in both C++ and Python."`.

**Comment by nverke** [comment]
Here is an example of the logits from a prompt that ended up choosing the wrong logit. 

```
Num tokens: 9967
[15.0703125  8.65625    8.15625  ] 1
[13.46875 12.53125 11.9375 ] 3577
[10.34375   10.203125   9.4765625] 1667
[11.09375   10.2890625  9.8359375] 2

package main
```

**Comment by masahi** [comment]
@nverke Are you sure that an input < 4k works fine with codellama 34B?

**Comment by nverke** [comment]
I think we are still working on exactly why this resolved the issue but my understanding is that the accuracy regression was more often causing issues with larger prompts due to compounding of the numerical differences. However we did see some examples of <4k token prompts causing the same issue.

**Comment by zxybazh** [comment]
A temporary solution is to use the `--no_cutlass_norm` option during quantization seems to avoid this problem on A100 GPU. Would be great to see if such fix also work for Metal platform to avoid NaN. CC @MasterJH5574

**Comment by zxybazh** [comment]
I can have a try for that, stay tuned.

**Comment by zxybazh** [comment]
I can confirm after fixing the eps, the issue is fixed on my testcase.

**Comment by masahi** [comment]
Thanks @zxybazh! I'll send fix PRs to cutlass and TVM.
